async_producer.go

```diff
@@ -840,6 +840,15 @@ func (bp *brokerProducer) run() {
 				continue
 			}
 
+			if msg.flags&fin == fin {
+				// New broker producer that was caught up by the retry loop
+				bp.parent.retryMessage(msg, ErrShuttingDown)
+				delete(bp.currentRetries[msg.Topic], msg.Partition)
```

Comment:
I'm not sure those 3 lines are necessary because the `bp.currentRetries` should be "empty" for that particular topic partition because `bp.needsRetry(msg)` returned `nil` above:
https://github.com/Shopify/sarama/blob/bad67e5b089437bc73f8034a95017e77be71e8b0/async_producer.go#L903-L909

Reply:
This is true that `bp.currentRetries[msg.Topic]` is `nil` in this case (`bp.currentRetries[msg.Topic][msg.Partition]` is set to `nil` in the "syn" case above). 
I decided to `delete` the key to ensure that the key does not exist to avoid confusion "key exist but nil" vs. "key does not exist" to mimic the other block of code above related to handling `fin` message.

TBH I'm not entirely sure about the purpose of the following code in handling `syn` case:
```
				if bp.currentRetries[msg.Topic] == nil {
					bp.currentRetries[msg.Topic] = make(map[int32]error)
				}
				bp.currentRetries[msg.Topic][msg.Partition] = nil
```

I don't see where it might be useful (for now).

Reply:
> TBH I'm not entirely sure about the purpose of the following code in handling `syn` case:
> 
> ```
> 				if bp.currentRetries[msg.Topic] == nil {
> 					bp.currentRetries[msg.Topic] = make(map[int32]error)
> 				}
> 				bp.currentRetries[msg.Topic][msg.Partition] = nil
> ```
> 
> I don't see where it might be useful (for now).

I believe when a `syn` message is seen the `brokerProducer` should:
- initialize `currentRetries` to prevent a panic when accessing `bp.currentRetries[topic][partition]` (typically on first use of the `brokerProducer`).
- clear a previous retries "state" for such topic partition.

Reply:
Thanks @slaunay for the reply!
I don't see how this could lead to any panics 🤔
I've given another thought and I think it does not really matter (as far as I can tell) whether the retries "metadata" structure is cleared or not (I was trying to imagine what if the code is changed in the future so this branch is executed even if `bp.currentRetries[msg.Topic][msg.Partition]` is not `nil`). So I will update the code to minimise the changes.